### PR TITLE
Keeper memory optimization by not holding snapshot in memory

### PR DESCRIPTION
**Pull request by zhanglistar**
Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Don't hold the latest snapshot in memory, instead, reading the snapshot if needed, sequence reading is fast to 200+MBps even on HDD using mmap system call.
Writing snapshot data directly to disk using compression method without holding original data and compressed data in memory.

Effect: about 20% saving memory
client command: `./utils/keeper-bench/keeper-bench --host=127.0.0.1:12183 -c 128 --generator=create_small_data`
new version:
image

master version(34092fa2a31dadff5efbd301e7d5582c925ab916):
image

### CODE REVIEW

File: src/Coordination/KeeperStateMachine.cpp [hunk 1]
```diff
@@ -308,7 +311,25 @@ int KeeperStateMachine::read_logical_snp_obj(
                             s.get_last_log_idx(), latest_snapshot_meta->get_last_log_idx());
             return -1;
         }
-        data_out = nuraft::buffer::clone(*latest_snapshot_buf);
+        int fd = ::open(latest_snapshot_path.c_str(), O_RDONLY);
```
Comment (alesapin) [inline src/Coordination/KeeperStateMachine.cpp:314]: Let's move this code into separate function.

File: src/Coordination/KeeperStateMachine.cpp [hunk 2]
```diff
@@ -308,7 +311,25 @@ int KeeperStateMachine::read_logical_snp_obj(
...
+        LOG_INFO(log, "Opening file {} for read_logical_snp_obj", latest_snapshot_path);
+        if (fd < 0)
+        {
+            LOG_WARNING(log, "Error opening {}.", latest_snapshot_path);
+            return -1;
+        }
+        auto file_size = ::lseek(fd, 0, SEEK_END);
```
Comment (alesapin) [inline src/Coordination/KeeperStateMachine.cpp:321]: Maybe we can use this code https://github.com/ClickHouse/ClickHouse/blob/e1fb220bb721a5db8b125c6768aa905d637084fe/src/Common/getMappedArea.h?

File: src/Coordination/KeeperStateMachine.cpp [hunk 3]
```diff
@@ -308,7 +311,25 @@ int KeeperStateMachine::read_logical_snp_obj(
...
+        ::lseek(fd, 0, SEEK_SET);
+        auto* chunk = reinterpret_cast<nuraft::byte*>(::mmap(nullptr, file_size, PROT_READ, MAP_FILE | MAP_SHARED, fd, 0));
```
Comment (alesapin) [inline src/Coordination/KeeperStateMachine.cpp:323]:
```suggestion
        auto * chunk = reinterpret_cast<nuraft::byte *>(::mmap(nullptr, file_size, PROT_READ, MAP_FILE | MAP_SHARED, fd, 0));
```

File: src/Coordination/KeeperStateMachine.cpp [hunk 4]
```diff
@@ -203,12 +204,15 @@ void KeeperStateMachine::create_snapshot(
         {
             {   /// Read storage data without locks and create snapshot
                 std::lock_guard lock(snapshots_lock);
-                auto snapshot_buf = snapshot_manager.serializeSnapshotToBuffer(*snapshot);
-                auto result_path = snapshot_manager.serializeSnapshotBufferToDisk(*snapshot_buf, snapshot->snapshot_meta->get_last_log_idx());
-                latest_snapshot_buf = snapshot_buf;
+                auto [path, error_code]= snapshot_manager.serializeSnapshotToDisk(*snapshot);
+                if (error_code)
+                {
+                    throw Exception(ErrorCodes::LOGICAL_ERROR, "Snapshot {} was created failed, error: {}",
```
Comment (alesapin) [inline src/Coordination/KeeperStateMachine.cpp:210]:
```suggestion
                    throw Exception(ErrorCodes::LOGICAL_ERROR, "Failed to create snapshot {} , error code: {}",
```
BTW, why is it `LOGICAL_ERROR`?

File: src/Coordination/KeeperSnapshotManager.h [hunk 5]
```diff
@@ -126,6 +130,13 @@ class KeeperSnapshotManager
         return 0;
     }
 
+    std::string getLatestSnapshotPath() const
+    {
+        if (!existing_snapshots.empty())
+            return existing_snapshots.at(getLatestSnapshotIndex());
```
Comment (alesapin) [inline src/Coordination/KeeperSnapshotManager.h:136]: Maybe check that we have it -- will get more clear error in case of bugs.

**Review by alesapin — COMMENTED** [review]
In general LGTM.

File: src/Coordination/KeeperStateMachine.cpp (hunk 4 above)
Comment (zhanglistar) [reply src/Coordination/KeeperStateMachine.cpp:210]: The directory should be writeable or enough space, but error happens, so I thought it was a LOGIC_ERROR.

Comment (zhanglistar) [reply src/Coordination/KeeperStateMachine.cpp:210]: What's your advice?

File: src/Coordination/KeeperSnapshotManager.h (hunk 5 above)
Comment (zhanglistar) [reply src/Coordination/KeeperSnapshotManager.h:136]: Yes, I will fix.

File: src/Coordination/KeeperStateMachine.cpp (hunk 4 above)
Comment (zhanglistar) [reply src/Coordination/KeeperStateMachine.cpp:210]: Should be IO error, I will change.

File: src/Coordination/KeeperStateMachine.cpp (hunk 2 above)
Comment (zhanglistar) [reply src/Coordination/KeeperStateMachine.cpp:321]: This is used only on Linux, not suitable for mac os x.

File: src/Coordination/KeeperStateMachine.cpp (hunk 1 above)
Comment (zhanglistar) [reply src/Coordination/KeeperStateMachine.cpp]: Done.

File: src/Coordination/KeeperStateMachine.cpp [hunk 6]
```diff
@@ -308,7 +342,11 @@ int KeeperStateMachine::read_logical_snp_obj(
                             s.get_last_log_idx(), latest_snapshot_meta->get_last_log_idx());
             return -1;
         }
-        data_out = nuraft::buffer::clone(*latest_snapshot_buf);
+        if (bufferFromFile(log, latest_snapshot_path, data_out))
```
Comment (alesapin) [inline src/Coordination/KeeperStateMachine.cpp:345]: Print errno?

Comment (zhanglistar) [reply src/Coordination/KeeperStateMachine.cpp:345]: errno has been printed in the inner function.

**Comment by alesapin** [comment]
`test_keeper_restore_from_snapshot/test.py::test_recover_from_snapshot` failure looks related to changes

**Comment by zhanglistar** [comment]
> `test_keeper_restore_from_snapshot/test.py::test_recover_from_snapshot` failure looks related to changes. @zhanglistar Can you run this integration test locally?

@alesapin Yes, I fix it now.

**Comment by zhanglistar** [comment]
> `test_keeper_restore_from_snapshot/test.py::test_recover_from_snapshot` failure looks related to changes. @zhanglistar Can you run this integration test locally?

@alesapin Could paste your ifconfig result of docker0? We have local environment problem when running integration tests  locally. It seems docker network related. Thanks.

File: src/Coordination/KeeperStateMachine.cpp [hunk 7]
```diff
@@ -281,6 +285,37 @@ void KeeperStateMachine::save_logical_snp_obj(
...
+    int fd = ::open(path.c_str(), O_RDONLY);
+    LOG_INFO(log, "Opening file {} for read_logical_snp_obj", path);
+    if (fd < 0)
+    {
+        LOG_WARNING(log, "Error opening {}, error: {}, errno: {}", path, std::strerror(errno), errno);
+        return errno;
+    }
+    auto file_size = ::lseek(fd, 0, SEEK_END);
+    ::lseek(fd, 0, SEEK_SET);
+    auto* chunk = reinterpret_cast<nuraft::byte*>(::mmap(nullptr, file_size, PROT_READ, MAP_FILE | MAP_SHARED, fd, 0));
```
Comment (alesapin) [inline src/Coordination/KeeperStateMachine.cpp:305]:
```suggestion
    auto * chunk = reinterpret_cast<nuraft::byte *>(::mmap(nullptr, file_size, PROT_READ, MAP_FILE | MAP_SHARED, fd, 0));
```

**Review by alesapin — APPROVED** [review]

**Comment by zhanglistar** [comment]
@alesapin Thanks for your review, can merge now?

**Comment by alesapin** [comment]
image
How is it possible not to merge?)

**Pull request merged**